trade.lic

```diff
@@ -0,0 +1,258 @@
+    deposit_coins(500, Room.current.id) if contract.destination_town == @hometown
+  end
+
+  def deposit_coins(withdraw_amt, return_to_id)
+    return if wealth(@hometown) < withdraw_amt
+    
+    walk_to(@town_data[@hometown]['deposit']['id'])
+    case bput('deposit all', 'you drop all your', 'You hand the clerk some coins', "You don't have any", 'There is no teller here', 'reached the maximum balance I can permit')
+    when 'There is no teller here'
+      return
+    end
+    minimize_coins(withdraw_amt).each { |amount| withdraw_exact_amount?(amount, @hometown) }
+    walk_to(return_to_id)
+  end
+
+  def find_closest_id(entity)
+    rooms = @town_data.to_h.values.select { |data| data[entity] }
+                             .map{ |data| data[entity]['id'] }
+    echo "rooms: #{rooms}" if $debug
+    return sort_destinations(rooms).first
+  end
+
+  def parse_contract(name=nil)
+    destination_re = /The guild office at (?:The )?(?<destination>.*) requires .*/
+    payment_re = /You estimate these goods are currently worth (.*) (Kronars|Dokoras|Lirums) on delivery\./
+    origin_re = /Trading Contract Issued by:  (?:The )?(?<origin>.*)/
+    contract = OpenStruct.new
+    contract.presented = true
+    contract.expired = false
+    fput("read my #{name} contract")
+    while line = get
+      echo line if $debug
+      if line =~ destination_re
+        contract.destination_town = line.match(destination_re).captures.first
+        contract.destination_outpost = @town_data[contract.destination_town]['trader_outpost']['id']
+        contract.destination_clerk = @town_data[contract.destination_town]['shipment_clerk']['id']
+        contract.destination_minister = @town_data[contract.destination_town]['trade_minister']['id']
+      elsif line =~ payment_re
+        # This is the last line of the contract so break when finished
+        contract.payment, contract.currency = line.match(payment_re).captures
+        break
+      elsif line =~ origin_re
+        contract.origin_town = line.match(origin_re).captures.first
+        contract.origin_clerk = @town_data[contract.origin_town]['shipment_clerk']['id']
+        contract.origin_minister = @town_data[contract.origin_town]['trade_minister']['id']
+        contract.origin_outpost = @town_data[contract.origin_town]['trader_outpost']['id']
+      elsif line =~ /This contract has yet to be/
+        contract.presented = false
+      elsif line =~ /The contract is now useless since it has expired\./
+        contract.expired = true
+        break
+      end
+    end
+    if $debug
+      echo "Contract origin: #{contract.origin_town}"
+      echo "Contract origin clerk room: #{contract.origin_clerk}"
+      echo "Contract origin minister room: #{contract.origin_minister}"
+      echo "Contract destination town: #{contract.destination_town}"
+      echo "Contracted presented to clerk? #{contract.presented}"
+    end
+    return contract
+  end
+
+  def recall_caravan
```

Comment:
It feels like this should return some sort of Caravan hash (similar to how `parse_contract` works). What do you think?

Reply:
I agree, then i can get rid of most of the instance variables